orttraining/orttraining/core/graph/optimizer_graph_builder.cc

```diff
@@ -443,17 +466,26 @@ Status OptimizerGraphBuilder::BuildInternal(
   // check if all gradients are finite
   ArgDef global_grad_norm_argdef;
   ArgDef global_grad_norm_finite_argdef;
-  if (opt_graph_config_.use_mixed_precision) {
-    //gradient norm for bfloat is not ready yet. skip it to unblock the testing
+
+  if (opt_graph_config_.enable_grad_norm_clip ||
+      (opt_graph_config_.use_mixed_precision &&
+       opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16)) {
+    //gradient norm for bfloat16 is not ready yet. skip it to unblock the testing
     //will add it back when it is ready
-    if (opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
-      ORT_RETURN_IF_ERROR(AddGradientNorm(
-          nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
-      ORT_RETURN_IF_ERROR(AddFiniteGradientCheck(
-          nodearg_name_generator, {global_grad_norm_argdef}, graph_defs, global_grad_norm_finite_argdef));
-      optimizer_graph_outputs[OptimizerOutputKey::GradientAllIsFinite] = global_grad_norm_finite_argdef.name;
-    }
+    ORT_RETURN_IF_ERROR(AddGradientNorm(
+        nodearg_name_generator, gradient_argdefs, graph_defs, global_grad_norm_argdef));
+    optimizer_graph_outputs[OptimizerOutputKey::GlobalGradientNorm] = global_grad_norm_argdef.name;
+  }
+
+  if (opt_graph_config_.use_mixed_precision &&
+      opt_graph_config_.mixed_precision_type == MixedPrecisionDataType::FP16) {
```

Comment:
I think this should be controlled by one single flag `check_gradient_magnitude`. This flag can be determined in ConfigureForTraining.

Reply:
I think this is a temporary condition as BF16 is not ready.

Reply:
I understand. It's a temporal condition which will be removed. If we determine `check_gradient_mmagnitude` in ConfigureForTraining and pass `check_gradient_mmagnitude` into this place, this file will have no change even when BF16 is done. If we spread such a condition around this file, it will be error-prone when we remove it.

---
In reply to: [520932982](https://github.com/microsoft/onnxruntime/pull/5728#discussion_r520932982) [](ancestors = 520932982)

Reply:
refactored.